crates/networking/rpc/engine/fork_choice.rs

```diff
@@ -211,58 +210,46 @@ async fn handle_forkchoice(
         .get_latest_valid_ancestor(fork_choice_state.head_block_hash)?
     {
-        warn!(
-            "Invalid fork choice state. Reason: Invalid ancestor {:#x}",
-            latest_valid_hash
-        );
-        Err(InvalidForkChoice::InvalidAncestor(latest_valid_hash))
-    } else {
-        // Check parent block hash in invalid_ancestors (if head block exists)
-        let check_parent = context
-            .storage
-            .get_block_header_by_hash(fork_choice_state.head_block_hash)?
-            .and_then(|head_block| {
-                debug!(
-                    "Checking parent for invalid ancestor {}",
-                    head_block.parent_hash
-                );
-                context
-                    .storage
-                    .get_latest_valid_ancestor(head_block.parent_hash)
-                    .ok()?
-            });
+        return Ok((
+            None,
+            ForkChoiceResponse::from(PayloadStatus::invalid_with(
+                latest_valid_hash,
+                InvalidForkChoice::InvalidAncestor(latest_valid_hash).to_string(),
+            )),
+        ));
+    }
 
-        // Check head block hash in invalid_ancestors
-        if let Some(latest_valid_hash) = check_parent {
-            Err(InvalidForkChoice::InvalidAncestor(latest_valid_hash))
-        } else {
-            // Check if there is an ongoing sync before applying the forkchoice
-            match context.syncer.status()? {
-                // Apply current fork choice
-                SyncStatus::Inactive => {
-                    // All checks passed, apply fork choice
-                    apply_fork_choice(
-                        &context.storage,
-                        fork_choice_state.head_block_hash,
-                        fork_choice_state.safe_block_hash,
-                        fork_choice_state.finalized_block_hash,
-                    )
-                    .await
-                }
-                // Restart sync if needed
-                _ => Err(InvalidForkChoice::Syncing),
-            }
+    // Check parent block hash in invalid_ancestors (if head block exists)
+    if let Some(head_block) = context
+        .storage
+        .get_block_header_by_hash(fork_choice_state.head_block_hash)?
+    {
+        if let Some(latest_valid_hash) = context
+            .storage
+            .get_latest_valid_ancestor(head_block.parent_hash)?
+        {
+            return Ok((
+                None,
+                ForkChoiceResponse::from(PayloadStatus::invalid_with(
+                    latest_valid_hash,
+                    InvalidForkChoice::InvalidAncestor(latest_valid_hash).to_string(),
+                )),
+            ));
         }
-    };
+    }
 
-    match fork_choice_res {
+    match apply_fork_choice(
+        &context.storage,
+        fork_choice_state.head_block_hash,
+        fork_choice_state.safe_block_hash,
+        fork_choice_state.finalized_block_hash,
+    )
+    .await
+    {
```

Comment:
We shouldn't apply the fork choice if we are currently syncing, this can lead to false positives if snap sync is enabled

Reply:
During an active snap sync we cannot rely on the current state (any block we have in the store may not have its state available), we cannot fetch a block from the store and set it as head if we have no state for it. Essentially, during a snap sync we "have no state"

Reply:
https://github.com/lambdaclass/ethrex/pull/2426/commits/f0474baa331b5868d6a9e4a43ee2a079a7bb09eb#diff-4fb4a7713efe0d060b54a7903029e99d7d0ecde127b62cfbacaae9344ce827e0R57-R60

What do you think?

Reply:
Ok, added a check specifically for snap sync.